ldap2pg/validators.py

```diff
@@ -21,8 +37,50 @@ def ldapquery(value):
     if 'filter' in query:
         query['filter'] = query['filter'].rstrip('\r\n')
 
-    # Clean value from old manual attribute
-    query.pop('attribute', None)
+    # Accept manual attributes, for legacy. Attribute inference should in all
+    # cases. Also, join attributes as predefined this way.
+    strlist_alias(query, 'attributes', 'attribute')
+    alias(query, 'joins', 'join')
+    query.setdefault('joins', {})
+
+    # ATTRIBUTES AND JOINS INFERENCE
+
+    legacy_attrs = query.get('attributes', [])
+    format_fields = (
+        list(format_fields or []) +
+        [a.split('.') for a in legacy_attrs])
+
+    # Now, loop format fields to list every attributes we need to query from
+    # LDAP directory, and for .sub.attr, detect whether we need a subquery (aka
+    # join).
+    attrs = set()
+    for field in format_fields:
+        attr, subattr = field[0], field[1:]
+        attrs.add(attr)
+        if not subattr:
```

Comment:
This clean up can cause some issues when both the raw attribute and a referenced sub-attribute is requested. Example: when both `{member.sAMAccountName}` and `{member}` are referenced, the additional LDAP query may not be performed and `sAMAccountName` won't be known.

Unit test:
```python
    v = ldapquery(
        dict(
            base='o=acme',
            join=dict(
                member=dict(filter='(objectClass=person)'))),
        format_fields=[
            ('sAMAccountName',),
            ('member', 'sAMAccountName'),
            ('member',),
        ]
    )

    assert 'member' in v['attributes']
    assert 'sAMAccountName' in v['attributes']
    assert 'member' in v['joins']
    assert ['sAMAccountName'] == v['joins']['member']['attributes']
```

Not sure if this is likely to happen.

When attributes are not defined for the join it will be removed below (in the “Post process joins” loop).

Reply:
good point. (finally) resolved.